### PR TITLE
Preliminary support for SGI IRIX

### PR DESCRIPTION
**Pull request by Xav101**
This is a series of patches to get the Xash3D dedicated server compiling on SGI IRIX 6.5.22 through 6.5.30. The next steps will be to improve engine support to get the server functional.

### CODE REVIEW

File: engine/common/whereami.c [hunk 1]
```diff
@@ -580,7 +580,7 @@ int WAI_PREFIX(getModulePath)(char* out, int capacity, int* dirname_length)
 }
 
 #elif defined(__DragonFly__) || defined(__FreeBSD__) || \
-      defined(__FreeBSD_kernel__) || defined(__NetBSD__) || defined(__OpenBSD__)
+      defined(__FreeBSD_kernel__) || defined(__NetBSD__) || defined(__OpenBSD_)
```
Comment (a1batross) [inline engine/common/whereami.c:583]: `__OpenBSD_`?..

File: engine/platform/posix/lib_posix.c [hunk 2]
```diff
@@ -16,6 +16,10 @@ GNU General Public License for more details.
 #include "platform/platform.h"
 #if XASH_LIB == LIB_POSIX
 #include <dlfcn.h>
+#ifdef XASH_IRIX
+#pragma message "Included"
```
Comment (a1batross) [inline engine/platform/posix/lib_posix.c:20]: Unneeded pragma

File: engine/platform/irix/dladdr.c [hunk 3]
```diff
@@ -0,0 +1,26 @@
+/*
```
Comment (a1batross) [inline engine/platform/irix/dladdr.c:1]: Can you add copyright/GPL notice and short description? You can copypaste it from other files. Like this: https://github.com/FWGS/xash3d-fwgs/blob/master/engine/common/zone.c#L1

File: engine/platform/posix/sys_posix.c [hunk 4]
```diff
@@ -158,9 +158,11 @@ void Platform_Shutdown( void ) {}
 double Platform_DoubleTime( void )
 {
 	struct timespec ts;
-
+	#if XASH_IRIX
```
Comment (a1batross) [inline engine/platform/posix/sys_posix.c:161]: Unneeded indentation for #if/#else/#endif.

Also, please add the space after opening and before closing parentheses everywhere. Consecutive parentheses must not have spaces in between.

While style rules wasn't followed everywhere, it's better good to slowly put in order such lines.

File: public/build.c [hunk 5]
```diff
@@ -107,6 +107,8 @@ const char *Q_PlatformStringByID( const int platform )
 		return "haiku";
 	case PLATFORM_SERENITY:
 		return "serenity";
+    case PLATFORM_IRIX:
```
Comment (a1batross) [inline public/build.c:110]: Use tabs for indentation.

File: engine/wscript [hunk 6]
```diff
@@ -68,7 +68,7 @@ def configure(conf):
 	if conf.env.DEST_OS == 'haiku':
 		conf.env.LIB_HAIKU = ['network']
 		conf.env.LIBPATH_HAIKU = ['/boot/system/lib']
-
+
```
Comment (a1batross) [inline engine/wscript:71]: Unneeded change.

File: public/build.h [hunk 7]
```diff
@@ -106,6 +107,8 @@ Then you can use another oneliner to query all variables:
 		#define XASH_NETBSD 1
 	#elif defined __OpenBSD__
 		#define XASH_OPENBSD 1
+    #elif defined __sgi
```
Comment (a1batross) [inline public/build.h:110]: Same, use tabs for indentation.

File: public/buildenums.h [hunk 8]
```diff
@@ -61,6 +62,8 @@ GNU General Public License for more details.
 	#define XASH_PLATFORM PLATFORM_HAIKU
 #elif XASH_SERENITY
 	#define XASH_PLATFORM PLATFORM_SERENITY
+#elif XASH_IRIX
+    #define XASH_PLATFORM PLATFORM_IRIX
```
Comment (a1batross) [inline public/buildenums.h:66]: Same, use tabs for indentation.

File: public/xash3d_mathlib.h [hunk 9]
```diff
@@ -77,6 +77,9 @@ GNU General Public License for more details.
 #define Q_round( x, y )	(floor( x / y + 0.5f ) * y )
 #define Q_rint(x)		((x) < 0.0f ? ((int)((x)-0.5f)) : ((int)((x)+0.5f)))
 
+#ifdef XASH_IRIX
+#undef isnan
```
Comment (a1batross) [inline public/xash3d_mathlib.h:81]: Is there something wrong about `isnan` on IRIX?

File: wscript [hunk 10]
```diff
@@ -160,9 +160,9 @@ def configure(conf):
 		conf.options.GL     = False
 	elif conf.env.MAGX:
 		conf.options.USE_SELECT       = True
-		conf.options.SDL12            = True
-		conf.options.NO_VGUI          = True
-		conf.options.GL               = False
+		conf.options.SDL12	      = True
```
Comment (a1batross) [inline wscript:163]: Unneeded changes.

File: wscript [hunk 11]
```diff
@@ -202,7 +202,7 @@ def configure(conf):
 		'-Werror=packed',
 		'-Werror=packed-not-aligned',
 		'-Wuninitialized', # older GCC versions have -Wmaybe-uninitialized enabled by this switch, which is not accurate
-                                   # so just warn, not error
+				   # so just warn, not error
```
Comment (a1batross) [inline wscript:205]: Unneeded changes.

File: wscript [hunk 12]
```diff
@@ -244,6 +244,14 @@ def configure(conf):
 		cxxflags += conf.filter_cxxflags(compiler_optional_flags, cflags)
 		cflags += conf.filter_cflags(compiler_optional_flags + c_compiler_optional_flags, cflags)
 
+	if conf.env.DEST_OS == 'irix' and conf.env.COMPILER_CC == 'gcc':
+		linkflags.remove('-Wl,--no-undefined')
+		linkflags.append('-Wl,--unresolved-symbols=ignore-all')
+		# check if we're in a sgug environment
+		if 'sgug' in os.environ['LD_LIBRARYN32_PATH']:
+			print("sgug")
```
Comment (a1batross) [inline wscript:252]: sgug? :D

File: wscript [hunk 13]
```diff
@@ -254,10 +262,10 @@ def configure(conf):
 	conf.env.SINGLE_BINARY = conf.options.SINGLE_BINARY or conf.env.DEDICATED
 
 	conf.env.NANOGL = conf.options.NANOGL or conf.options.ALL_RENDERERS
-	conf.env.GLWES  = conf.options.GLWES or conf.options.ALL_RENDERERS
-	conf.env.GL4ES  = conf.options.GL4ES or conf.options.ALL_RENDERERS
-	conf.env.GL     = conf.options.GL or conf.options.ALL_RENDERERS
-	conf.env.SOFT   = conf.options.SOFT or conf.options.ALL_RENDERERS
+	conf.env.GLWES	= conf.options.GLWES or conf.options.ALL_RENDERERS
```
Comment (a1batross) [inline wscript:265]: Unneeded changes.

File: wscript [hunk 14]
```diff
@@ -327,7 +335,7 @@ int main(void) { return 0; }''',
 	if conf.env.DEST_OS != 'win32':
 		strcasestr_frag = '''#include <string.h>
 int main(int argc, char **argv) { strcasestr(argv[1], argv[2]); return 0; }'''
-		strchrnul_frag  = '''#include <string.h>
+		strchrnul_frag	= '''#include <string.h>
 int main(int argc, char **argv) { strchrnul(argv[1], 'x'); return 0; }'''
```
Comment (a1batross) [inline wscript:339]: Unneeded changes.

File: xash-exec [hunk 15]
```diff
@@ -0,0 +1,7 @@
+#!/usr/sgug/bin/bash
```
Comment (a1batross) [inline xash-exec:1]: It seems to be IRIX specific?

Could you move it to engine/platform/irix then?

**Comment by a1batross** [comment]
In general, I don't see anything critical in these patches. It's all very much bare-bones IRIX support, with platform-specific code here and there.

Please, just remove unneeded changes. Also, it would be nice to add the tags to the commit messages.

File: public/xash3d_mathlib.h (hunk 9 above)
Comment (Xav101) [reply public/xash3d_mathlib.h:81]: The default system `isnan` seems to be broken in some cases. In one specific instance it causes the build to fail. The GCC 9.2 toolchain for IRIX is supposed to prefer a fixed version but for some reason this seems to not work on my Octane2 running IRIX 6.5.30 but did work on my other machine running 6.5.22. However I haven't been able to figure out the problem yet so this serves as fix for the moment.

File: common/xash3d_types.h [hunk 16]
```diff
@@ -120,7 +120,7 @@ typedef uint64_t longtime_t;
 #define LittleLongSW(x) (x = LittleLong(x) )
 #define LittleShort(x) ((short)( (((short)(x) >> 8) & 255) + (((short)(x) & 255) << 8)))
 #define LittleShortSW(x) (x = LittleShort(x) )
-_inline float LittleFloat( float f )
+static inline float LittleFloat( float f )
```
Comment (nekonomicon) [inline common/xash3d_types.h:123]: You can include `port.h` where needed:
https://github.com/FWGS/xash3d-fwgs/blob/master/common/port.h#L37=

**Pull request merged**